design-proposals/preferences.md

```diff
@@ -0,0 +1,584 @@
+
+This document proposes an extension to the existing [VirtualMachineFlavor](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [VirtualMachineClusterFlavor](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor) abstractions for KubeVirt called Virtual Machine Preferences to bring KubeVirt’s VM creation workflows into alignment with prior art.
+
+## Problem Statement
+
+[VirtualMachineFlavor](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [VirtualMachineClusterFlavor](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor) were  introduced by [6dc548459f](https://github.com/kubevirt/kubevirt/commit/6dc548459fa17d7f9601cbc251088d2b70a2a96a) with the current [VirtualMachineFlavorProfile](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) definition only containing support to express the [CPU](http://kubevirt.io/api-reference/main/definitions.html#_v1_cpu) element of a [DomainSpec](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) that is currently applied in full to a `VMI`.
+
+While working to extend this implementation it was noted that there was no easy way of providing preferences directly within [VirtualMachineFlavorProfile](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) without causing the duplication of said profiles for every combination of resource size and set of preferences being defined.
+
+Additionally the preferences would be tightly coupled with the sizing of resources within the flavor. As a result third parties such as application or distribution vendors would not be able to provided generic preferences independently.
+
+In OpenStack for example this issue is mostly avoided through the use of [image properties](https://docs.openstack.org/glance/latest/admin/useful-image-properties.html). This is where the boot image of an instance (`VMI`) is decorated with properties in order to override or augment preferences defined by [flavors](https://docs.openstack.org/nova/latest/user/flavors.html).
+
+As this isn't possible within `KubeVirt` another mechanism is required to separate the quantitative compute resources defined by flavors and qualitative preferences we also want to provide.
+
+## User Stories
+
+* As an operator or user I would like to provide flavors that define the compute resources available to a VM.
+
+* As an operator or user I would like to provide preferences that relate to the performance characteristics of a VM within these flavors.
```

Comment:
In which way they relate?

Reply:
See the examples at the end where I suggest two options for providing the preferences, either embedded in the flavor or alongside within the VM definition.